core/corehttp/gateway_handler.go

```diff
@@ -538,12 +537,23 @@ func (i *gatewayHandler) serveCBOR(w http.ResponseWriter, r *http.Request, n for
 	}
 	modtime := time.Unix(1, 0)
 
-	b, err := json.Marshal(n)
-	if err != nil {
-		internalWebError(w, err)
-		return
+	var contentType string
+	var data []byte
+	if r.URL.Query().Get("enc") == "json" || r.Header.Get("Content-Type") == "application/json" {
```

Comment:
I think this should be: `r.Header.Get("Accept")`
That header in the request is about the body of the request.

Reply:
... oh, except "Accept" can be a comma-separated list, so the header value needs to be parsed - not just `==` compared.

Reply:
this seems to work pretty well:

```
diff --git a/core/corehttp/gateway_handler.go b/core/corehttp/gateway_handler.go
index 11e5db671..5b847ae10 100644
--- a/core/corehttp/gateway_handler.go
+++ b/core/corehttp/gateway_handler.go
@@ -39,6 +39,7 @@ const (
 )
 
 var onlyAscii = regexp.MustCompile("[[:^ascii:]]")
+var anyJSON = regexp.MustCompile(`^application/(|.+\+)json$`)
 
 // HTML-based redirect for errors which can be recovered from, but we want
 // to provide hint to people that they should fix things on their end.
@@ -537,9 +538,24 @@ func (i *gatewayHandler) serveCBOR(w http.ResponseWriter, r *http.Request, n for
 	}
 	modtime := time.Unix(1, 0)
 
+	useJSON := false
+	if r.URL.Query().Get("enc") == "json" {
+		useJSON = true
+	} else {
+		for _, v := range r.Header.Values("Accept") {
+			for _, a := range strings.Split(v, ",") {
+				mediatype, _, err := mime.ParseMediaType(a)
+				if err == nil && anyJSON.MatchString(mediatype) {
+					useJSON = true
+					break
+				}
+			}
+		}
+	}
+
 	var contentType string
 	var data []byte
-	if r.URL.Query().Get("enc") == "json" || r.Header.Get("Content-Type") == "application/json" {
+	if useJSON {
 		contentType = "application/json"
 		b, err := json.Marshal(n)
 		if err != nil {
```

Reply:
d'oh! : forgot the label on the outer loop so the break can break all the way out.